### ISSUE TITLE
salt 2016.3.0 will not start on FreeBSD

### ISSUE THREAD

**Issue by maschinetheist** [saltstack/salt#33807]
### Description of Issue/Question

salt 2016.3.0 does not start on FreeBSD 10.3. py27-salt-2016.3.0 was installed with defaults from FreeBSD ports. When I run /usr/local/etc/rc.d/salt_master onestart (or start or even salt-master -l debug), it just hangs. 

``` bash
$ sudo /usr/local/etc/rc.d/salt_master onestart
Starting salt_master.

```

When I truss the process, I get this:

``` bash
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
stat("/etc/nsswitch.conf",{ mode=-rw-r--r-- ,inode=12469,size=321,blksize=4096 }) = 0 (0x0)
```

This goes on and on and on and never stops.
### Setup

/usr/local/etc/salt/master (mostly defaults, here are the uncommented options):

```
$ egrep -v "^#|^$" master
interface: 0.0.0.0
user: root
worker_threads: 10
file_roots:
  base:
    - /srv/salt/states
pillar_roots:
  base:
    - /srv/salt/pillar
```
### Versions Report

```
$ salt-master --versions-report
Salt Version:
           Salt: 2016.3.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.11 (default, May 19 2016, 01:23:53)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.3
            ZMQ: 4.1.4

System Versions:
           dist:
        machine: amd64
        release: 10.3-RELEASE-p3
         system: FreeBSD
        version: Not Installed
```

**Comment by Ch3LL** [comment]
@maschinetheist I am not able to replicate this behavior even using your exact master configuration. Everything is starting up fine and i can communicate with a minion. Did you upgrade your packages or is this a clean install?

**Comment by maschinetheist** [comment]
Hi @Ch3LL, this is a completely clean installation of Salt and FreeBSD. System is less than 10 days old. System has been fully updated too. I can't think of a reason why it would be trying to stat or read /etc/nsswitch.conf but I do have an idea.

I started doing some more research on my working, clean FreeBSD 10.3 VM. On that VM, I was able to bring up salt-master successfully. While trussing the salt-master process, I noticed that it was trying to open /usr/local/lib/python2.7/site-packages/pysss.so which was missing on the working VM. Whether that was salt-master or some other module, I don't know but I'll try to find out. pysss.so belongs to sssd package/daemon which would explain nsswitch.conf but I'm not sure why anything would read it when the daemon is stopped. I then removed sssd package on my system that wouldn't start salt processes and now salt-master starts up successfully. Success!

What's strange is that sssd was not running on any of my systems. I'll dig deeper to see if maybe some other tool caused it to hiccup.

**Comment by rch317** [comment]
Just curious if there is a better solution than removing the sssd package.  That *did* work, as we were seeing the same problem as well.

**Comment by Ch3LL** [comment]
Okay thanks for bumping this for me @rch317  apologies for nto seeing this earlier @maschinetheist and thanks for the great investigation work.

Looks like it is this line [here](https://github.com/saltstack/salt/blob/develop/salt/utils/__init__.py#L2766) where its hanging.  Even if you open a python interpreter on freebsd:

```
>>> user='root'
>>> group_names = list(pysss.getgrouplist(user))
^C^C^Z
```
It just hangs....It looks like it would only get to this part in the code if you had pysss installed. 

getgrouplist exists:

```
>>> pysss.__dict__.keys()
['__file__', '__package__', 'getgrouplist', '__name__', 'password', 'local', '__doc__']
```

```
>>> help(pysss.getgrouplist)
Help on built-in function getgrouplist in module pysss:

getgrouplist(...)
    Get list of groups user belongs to.
    
    NOTE: The interface uses the system NSS calls and is not limited to users served by the SSSD!
    :param username: name of user to get list for
```

```
root@ch3ll-free:~ # cat /etc/nsswitch.conf
#
# nsswitch.conf(5) - name service switch configuration file
# $FreeBSD: releng/10.1/etc/nsswitch.conf 224765 2011-08-10 20:52:02Z dougb $
#
group: compat
group_compat: nis
hosts: files dns
networks: files
passwd: compat
passwd_compat: nis
shells: files
services: compat
services_compat: nis
protocols: files
rpc: files
```

Okay I think I found the issue and its with:

```
group: compat
group_compat: nis
```

In the nssswitch.conf file. I believe its looking for a nis service to be started and there is not (although my knowledge around nis is fuzzy)

If i change it to the following:

```
group: sss
group_compat: sss
```

and use the SSSD instead it works. I'm assuming if you had a nis service setup as well it would work as well.

**Comment by cachedout** [comment]
@kev009 Does this ring any bells for you by chance?

**Comment by amendlik** [comment]
I ran into this same problem, as well, and managed to get it working by building a custom patched version of the sssd package. Here is what I learned:

- There is a bug in `pysss.getgrouplist()` that causes it to hang on FreeBSD.
- The `pysss` Python module is provided by the `sssd` package
- A [patch](https://lists.fedorahosted.org/archives/list/sssd-devel@lists.fedorahosted.org/thread/VAQK7VJUNTVOI6XYDZXFCZHZKXU3GYCA/) was submitted to address this, but it was never applied.
- Running Salt under Python 3 avoids this problem because it uses `os.getgrouplist()` instead of `pysss.getgrouplist()`. See code [here](https://github.com/saltstack/salt/blob/7a4b15302016414fac3feed6921065dfc788b64d/salt/utils/__init__.py#L2957).

I used portshaker and poudriere to build a custom package for `sssd` by including the patch referenced above, and it works fine. I now have a working SSSD configuration and SaltStack running.

In my case, I was already building custom packages for SSSD to get FreeIPA support, so it was not that difficult to add a patch. I think there are a few options for a less labor-intensive solution for the masses.

1. SaltStack could publish this patched `sssd` in their [FreeBSD repo](https://repo.saltstack.com/freebsd/).
2. SaltStack could work with the maintainer of the `sssd` port to get the patch applied.
3. Switch to the Python 3 version of Salt

Let me know what else I can provide to help resolve this.
